bin/dist_get

```diff
@@ -0,0 +1,143 @@
+#!/bin/bash
+
+die() {
+	echo "$@" >&2
+	exit 1
+}
+
+have_binary() {
+	type "$1" > /dev/null
+}
+
+check_writeable() {
+	printf "" > "$1" && rm "$1"
+}
+
+download() {
+	local url="$1"
+	local output="$2"
+
+	if [ -z "$url" ] || [ -z "$output" ]; then
+		die "download takes exactly two arguments. was given '$@'"
+	fi
+
+	if ! check_writeable "$output"; then
+		die "download error: cannot write to $output"
+	fi
+
+	printf 'Downloading "%s" to "%s"\n' "$url" "$output"
+	if have_binary wget; then
+		wget "$url" -O "$output"
+	elif have_binary curl; then
+		curl --silent "$url" > "$output"
+	elif have_binary fetch; then
+		fetch "$url" -o "$output"
+	else
+		die "no binary found to download $url. exiting."
+	fi
+}
+
+unarchive() {
+	local archivetype="$1"
+	local infile="$2"
+	local outfile="$3"
+	local distname="$4"
+
+	if ! check_writeable "$outfile"; then
+		die "unarchive error: cannot write to $outfile"
+	fi
+
+	case $archivetype in
+		tar.gz)
+			if have_binary tar; then
+				echo "==> using 'tar' to extract binary from archive"
+				tar -x "$distname/$distname" -f "$infile" -O > "$outfile"
+			else
+				die "no binary on system for extracting tar files"
+			fi
+			;;
+		zip)
+			if have_binary unzip; then
+				echo "==> using 'unzip' to extract binary from archive"
+				unzip -p "$infile" "$distname/$distname" > "$outfile"
+			else
+				die "no installed method for extracting .zip archives"
+			fi
+			;;
+		*)
+			die "unrecognized archive type '$archivetype'"
+	esac
+
+	chmod +x "$outfile"
+}
+
+get_go_vars() {
+	if [ ! -z "$GOOS" ] && [ ! -z "$GOARCH" ]; then
+		printf "%s-%s" "$GOOS" "$GOARCH"
+	fi
+
+	if have_binary go; then
+		printf "%s-%s" "$(go env GOOS)" "$(go env GOARCH)"
+	else
+		die "no way of determining system GOOS and GOARCH\nPlease manually set GOOS and GOARCH then retry."
+	fi
+}
+
+mkurl() {
+	local name="$1"
+	local vers="$2"
+	local archive="$3"
+
+	local govars=$(get_go_vars)
```

Comment:
If go_get_vars() dies, then we will just have an empty "govars" and continue. I am not sure that is what we want.